### DISCUSSION TITLE
RLS policy with patched-column dependency

### DISCUSSION THREAD

**Discussion by scoover** [supabase/supabase#27333] [Questions]
Is it possible to create an UPDATE RLS policy that varies the authorization based on what fields are being patched?

Essentially I have a roles table called `user_team` where rows allow users access to teams, establish their permissions and have a `archived` bool. 

1. I'd like to allow users who are team admins (i.e. have a row with the team_id and that row is set to "admin") to be able to change any/all cells for any user within the team.
2. I'd like to be able to allow a non-admin user to modify just the `archived` cell for their own row (i.e. I do not want them to be able to change their role).

The policy below is not acceptable because it allows the user to change any/all cells within their row.

```
ALTER POLICY "UPDATE allowed by team admins"
ON "public"."user_team"
TO authenticated
USING (
  team_id IN (
    SELECT ut.team_id
    FROM user_team ut
    WHERE 
      (auth.jwt() ->> 'email'::text) = ut.email
      AND (
        ut.role = 'admin'::text 
        OR ut.email = auth.jwt() ->> 'email'::text
      )
  )
)
WITH CHECK (true);
```

Thanks.

P.S. ChatGPT proposed a CHECK that involved `NEW.* IS DISTINCT FROM OLD.*...` but Supabase doesn't seem to accept this in a policy.

**Comment by GaryAustin1** [comment]
NEW is not available in RLS policies so not sure why the AI would suggest that.

Using looks at the current row and With Check looks at the new row.  But you can't for instance in Using see the incoming data.

You can't control columns in RLS. 

You probably need to look at locking RLS for updates and then using an RPC call with a security definer function and your rules checked in the function and do the update that way.

**Comment by scoover** [comment]
Just so I understand… I would need USING (on the current table data) to see that the user has or doesn’t have the “admin” role and I would need CHECK to look at the incoming data to see that the user was bumping their own role permissions… but these two filters can’t resolve the problem because USING and CHECK can’t share information these key pieces of information with each other?

Thanks, Gary. Appreciate the quick response.

**Reply by GaryAustin1** [reply]
Normally you should not start a new comment by reply to the one you are responding back to.

You can check admin role in either as that is not part of the data in the row.

But you are correct that you cannot see the incoming data in relation to the old data in either USING or WITH CHECK.  They only see the row as it exists in the database and the row as it will be written to the database.

You also might be able to do what you want with a trigger function (I was not focused on the detail, but the RLS limits).  A trigger function can look at the NEW data and compare to the OLD data for update.  It can also do security with SQL as auth.jwt() works and you have access to other tables.  You can also see the current postgres role to bypass these checks for postgres user or the like.

**Reply by scoover** [reply]
Sounds promising @GaryAustin1! Will definitely look into a trigger function.

Best, Randy

P.S. Sorry about the _new_ comment. Just replied in my email client and thought it would do the right thing.
